File: trevrpc-go/wire.go
package trevrpc

import (
	"fmt"
	"strings"
)

const (
	// ALPN is the QUIC application protocol negotiated by TrevRPC peers.
	ALPN = "trevrpc/1"
	// WireVersion is the current TrevRPC wire protocol version.
	WireVersion = uint32(1)

	// MaxMetadataEntries is the maximum number of metadata entries per message.
	MaxMetadataEntries = 64
	// MaxMetadataKeyLen is the maximum metadata key length in bytes.
	MaxMetadataKeyLen = 128
	// MaxMetadataValueLen is the maximum metadata value length in bytes.
	MaxMetadataValueLen = 8 * 1024
	// MaxMetadataTotalSize is the maximum combined metadata key and value size in bytes.
	MaxMetadataTotalSize = 64 * 1024

	// ReservedMetadataPrefix is reserved for runtime-owned metadata keys.
	ReservedMetadataPrefix = "trevrpc-"
)

// Metadata stores request or response metadata as byte values keyed by normalized names.
type Metadata map[string][]byte

// RpcKind identifies whether an RPC is unary or streaming.
type RpcKind int32

const (
	RpcKindUnary                  RpcKind = 0
	RpcKindClientStreaming        RpcKind = 1
	RpcKindServerStreaming        RpcKind = 2
	RpcKindBidirectionalStreaming RpcKind = 3
)

// RpcStreamFrameKind identifies whether a streaming frame carries a message or terminal status.
type RpcStreamFrameKind int32

const (
	RpcStreamFrameKindMessage RpcStreamFrameKind = 0
	RpcStreamFrameKindStatus  RpcStreamFrameKind = 1
)

// RpcRequest is the wire representation of an RPC request.
type RpcRequest struct {
	Service      string   `protobuf:"bytes,1,opt,name=service,proto3" json:"service,omitempty"`
	Method       string   `protobuf:"bytes,2,opt,name=method,proto3" json:"method,omitempty"`
	Body         []byte   `protobuf:"bytes,3,opt,name=body,proto3" json:"body,omitempty"`
	Metadata     Metadata `protobuf:"bytes,4,rep,name=metadata,proto3" json:"metadata,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	Kind         RpcKind  `protobuf:"varint,5,opt,name=kind,proto3,enum=trevrpc.RpcKind" json:"kind,omitempty"`
	Version      uint32   `protobuf:"varint,6,opt,name=version,proto3" json:"version,omitempty"`
	TimeoutNanos uint64   `protobuf:"varint,7,opt,name=timeout_nanos,json=timeoutNanos,proto3" json:"timeout_nanos,omitempty"`
}

// NewRpcRequest creates a unary RPC request with the current wire version.
func NewRpcRequest(service, method string, body []byte) *RpcRequest {
	return &RpcRequest{
		Service:  service,
		Method:   method,
		Body:     body,
		Metadata: Metadata{},
		Kind:     RpcKindUnary,
		Version:  WireVersion,
	}
}

// Reset clears the request for protobuf compatibility.
func (m *RpcRequest) Reset() { *m = RpcRequest{} }

// String returns the request as a debug string for protobuf compatibility.
func (m *RpcRequest) String() string { return fmt.Sprintf("%+v", *m) }

// ProtoMessage marks RpcRequest as a protobuf message.
func (*RpcRequest) ProtoMessage() {}

// ValidateProtocol validates the request wire version and RPC kind.
func (m *RpcRequest) ValidateProtocol() error {
	if m.Version != WireVersion {
		return FailedPrecondition(fmt.Sprintf("unsupported TrevRPC wire version %d; expected %d", m.Version, WireVersion))
	}

	if !m.Kind.IsValid() {
		return InvalidArgument(fmt.Sprintf("unsupported TrevRPC RPC kind %d", m.Kind))
	}

	return nil
}

// IsValid reports whether k is a supported RPC kind.
func (k RpcKind) IsValid() bool {
	switch k {
	case RpcKindUnary, RpcKindClientStreaming, RpcKindServerStreaming, RpcKindBidirectionalStreaming:
		return true
	default:
		return false
	}
}

// RPCKind returns the request RPC kind, defaulting to unary for invalid values.
func (m *RpcRequest) RPCKind() RpcKind {
	if m.Kind.IsValid() {
		return m.Kind
	}

	return RpcKindUnary
}

// RpcResponse is the wire representation of an RPC response.
type RpcResponse struct {
	Status   uint32   `protobuf:"varint,1,opt,name=status,proto3" json:"status,omitempty"`
	Message  string   `protobuf:"bytes,2,opt,name=message,proto3" json:"message,omitempty"`
	Body     []byte   `protobuf:"bytes,3,opt,name=body,proto3" json:"body,omitempty"`
	Metadata Metadata `protobuf:"bytes,4,rep,name=metadata,proto3" json:"metadata,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
}

// OKResponse creates a successful RPC response with the provided body.
func OKResponse(body []byte) *RpcResponse {
	return OK().IntoResponse(body)
}

// Reset clears the response for protobuf compatibility.
func (m *RpcResponse) Reset() { *m = RpcResponse{} }

// String returns the response as a debug string for protobuf compatibility.
func (m *RpcResponse) String() string { return fmt.Sprintf("%+v", *m) }

// ProtoMessage marks RpcResponse as a protobuf message.
func (*RpcResponse) ProtoMessage() {}

// RpcStreamFrame is the wire representation of one streaming RPC frame.
type RpcStreamFrame struct {
	Kind     RpcStreamFrameKind `protobuf:"varint,1,opt,name=kind,proto3,enum=trevrpc.RpcStreamFrameKind" json:"kind,omitempty"`
	Status   uint32             `protobuf:"varint,2,opt,name=status,proto3" json:"status,omitempty"`
	Message  string             `protobuf:"bytes,3,opt,name=message,proto3" json:"message,omitempty"`
	Body     []byte             `protobuf:"bytes,4,opt,name=body,proto3" json:"body,omitempty"`
	Metadata Metadata           `protobuf:"bytes,5,rep,name=metadata,proto3" json:"metadata,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
}

// MessageFrame creates a stream message frame carrying a protobuf body.
func MessageFrame(body []byte) *RpcStreamFrame {
	return &RpcStreamFrame{
		Kind:     RpcStreamFrameKindMessage,
		Status:   uint32(CodeOK),
		Body:     body,
		Metadata: Metadata{},
	}
}

// StatusFrame creates a terminal status frame without metadata.
func StatusFrame(status *Status) *RpcStreamFrame {
	return StatusFrameWithMetadata(status, Metadata{})
}

// StatusFrameWithMetadata creates a terminal status frame with metadata.
func StatusFrameWithMetadata(status *Status, metadata Metadata) *RpcStreamFrame {
	if status == nil {
		status = OK()
	}

	return &RpcStreamFrame{
		Kind:     RpcStreamFrameKindStatus,
		Status:   uint32(status.Code),
		Message:  status.Message,
		Metadata: metadata,
	}
}

// Reset clears the stream frame for protobuf compatibility.
func (m *RpcStreamFrame) Reset() { *m = RpcStreamFrame{} }

// String returns the stream frame as a debug string for protobuf compatibility.
func (m *RpcStreamFrame) String() string { return fmt.Sprintf("%+v", *m) }

// ProtoMessage marks RpcStreamFrame as a protobuf message.
func (*RpcStreamFrame) ProtoMessage() {}

// FrameKind returns the decoded stream frame kind.
func (m *RpcStreamFrame) FrameKind() (RpcStreamFrameKind, bool) {
	switch m.Kind {
	case RpcStreamFrameKindMessage, RpcStreamFrameKindStatus:
		return m.Kind, true
	default:
		return 0, false
	}
}

// StatusValue returns the status represented by the frame status fields.
func (m *RpcStreamFrame) StatusValue() *Status {
	return NewStatus(CodeFromUint32(m.Status), m.Message)
}

// NormalizeMetadataKey normalizes a metadata key to lowercase ASCII.
func NormalizeMetadataKey(key string) string {
	return strings.ToLower(key)
}

// ValidateMetadata validates metadata key syntax, value sizes, and total metadata limits.
func ValidateMetadata(metadata Metadata) error {
	if len(metadata) > MaxMetadataEntries {
		return InvalidArgument(fmt.Sprintf("metadata has %d entries, maximum is %d", len(metadata), MaxMetadataEntries))
	}

	totalSize := 0
	for key, value := range metadata {
		if err := validateMetadataKey(key); err != nil {
			return err
		}

		if len(value) > MaxMetadataValueLen {
			return InvalidArgument(fmt.Sprintf("metadata value %q is %d bytes, maximum is %d", key, len(value), MaxMetadataValueLen))
		}

		totalSize += len(key) + len(value)
	}

	if totalSize > MaxMetadataTotalSize {
		return InvalidArgument(fmt.Sprintf("metadata is %d bytes, maximum is %d", totalSize, MaxMetadataTotalSize))
	}

	return nil
}

func validateMetadataKey(key string) error {
	if key == "" {
		return InvalidArgument("metadata key is empty")
	}

	if len(key) > MaxMetadataKeyLen {
		return InvalidArgument(fmt.Sprintf("metadata key %q is %d bytes, maximum is %d", key, len(key), MaxMetadataKeyLen))
	}

	if strings.HasPrefix(key, ReservedMetadataPrefix) {
		return InvalidArgument(fmt.Sprintf("metadata key %q uses reserved prefix %q", key, ReservedMetadataPrefix))
	}

	for i := 0; i < len(key); i++ {
		if !isMetadataKeyByte(key[i]) {
			return InvalidArgument(fmt.Sprintf("metadata key %q must use lowercase ASCII letters, digits, '.', '_' or '-'", key))
		}
	}

	return nil
}

func isMetadataKeyByte(value byte) bool {
	return (value >= 'a' && value <= 'z') || (value >= '0' && value <= '9') || value == '.' || value == '_' || value == '-'
}

File: src/wire.rs
use std::collections::HashMap;

pub type Metadata = HashMap<String, Vec<u8>>;

pub const MAX_METADATA_ENTRIES: usize = 64;
pub const MAX_METADATA_KEY_LEN: usize = 128;
pub const MAX_METADATA_VALUE_LEN: usize = 8 * 1024;
pub const MAX_METADATA_TOTAL_SIZE: usize = 64 * 1024;
pub const RESERVED_METADATA_PREFIX: &str = "trevrpc-";

#[must_use]
pub fn normalize_metadata_key(key: &str) -> String {
    key.to_ascii_lowercase()
}

pub fn validate_metadata(metadata: &Metadata) -> std::result::Result<(), crate::Status> {
    if metadata.len() > MAX_METADATA_ENTRIES {
        return Err(crate::Status::invalid_argument(format!(
            "metadata has {} entries, maximum is {MAX_METADATA_ENTRIES}",
            metadata.len()
        )));
    }

    let mut total_size = 0_usize;

    for (key, value) in metadata {
        validate_metadata_key(key)?;

        if value.len() > MAX_METADATA_VALUE_LEN {
            return Err(crate::Status::invalid_argument(format!(
                "metadata value {key:?} is {} bytes, maximum is {MAX_METADATA_VALUE_LEN}",
                value.len()
            )));
        }

        total_size = total_size
            .saturating_add(key.len())
            .saturating_add(value.len());
    }

    if total_size > MAX_METADATA_TOTAL_SIZE {
        return Err(crate::Status::invalid_argument(format!(
            "metadata is {total_size} bytes, maximum is {MAX_METADATA_TOTAL_SIZE}"
        )));
    }

    Ok(())
}

fn validate_metadata_key(key: &str) -> std::result::Result<(), crate::Status> {
    if key.is_empty() {
        return Err(crate::Status::invalid_argument("metadata key is empty"));
    }

    if key.len() > MAX_METADATA_KEY_LEN {
        return Err(crate::Status::invalid_argument(format!(
            "metadata key {key:?} is {} bytes, maximum is {MAX_METADATA_KEY_LEN}",
            key.len()
        )));
    }

    if key.starts_with(RESERVED_METADATA_PREFIX) {
        return Err(crate::Status::invalid_argument(format!(
            "metadata key {key:?} uses reserved prefix {RESERVED_METADATA_PREFIX:?}"
        )));
    }

    if !key.bytes().all(is_metadata_key_byte) {
        return Err(crate::Status::invalid_argument(format!(
            "metadata key {key:?} must use lowercase ASCII letters, digits, '.', '_' or '-'"
        )));
    }

    Ok(())
}

const fn is_metadata_key_byte(byte: u8) -> bool {
    matches!(byte, b'a'..=b'z' | b'0'..=b'9' | b'.' | b'_' | b'-')
}

#[derive(Clone, PartialEq, prost::Message)]
pub struct RpcRequest {
    #[prost(string, tag = "1")]
    pub service: String,
    #[prost(string, tag = "2")]
    pub method: String,
    #[prost(bytes = "vec", tag = "3")]
    pub body: Vec<u8>,
    #[prost(map = "string, bytes", tag = "4")]
    pub metadata: Metadata,
}

impl RpcRequest {
    #[must_use]
    pub fn new(service: impl Into<String>, method: impl Into<String>, body: Vec<u8>) -> Self {
        Self {
            service: service.into(),
            method: method.into(),
            body,
            metadata: Metadata::new(),
        }
    }

    #[must_use]
    pub fn with_metadata(mut self, metadata: Metadata) -> Self {
        self.metadata = metadata;
        self
    }
}

#[derive(Clone, PartialEq, prost::Message)]
pub struct RpcResponse {
    #[prost(uint32, tag = "1")]
    pub status: u32,
    #[prost(string, tag = "2")]
    pub message: String,
    #[prost(bytes = "vec", tag = "3")]
    pub body: Vec<u8>,
    #[prost(map = "string, bytes", tag = "4")]
    pub metadata: Metadata,
}

impl RpcResponse {
    #[must_use]
    pub fn ok(body: Vec<u8>) -> Self {
        crate::Status::ok().into_response(body)
    }
}

#[cfg(test)]
mod tests {
    use crate::{Code, Metadata};

    use super::{
        MAX_METADATA_ENTRIES, MAX_METADATA_TOTAL_SIZE, MAX_METADATA_VALUE_LEN,
        normalize_metadata_key, validate_metadata,
    };

    #[test]
    fn validates_well_formed_metadata() {
        let mut metadata = Metadata::new();
        metadata.insert("authorization".to_owned(), b"Bearer token".to_vec());

        assert_eq!(validate_metadata(&metadata), Ok(()));
    }

    #[test]
    fn normalizes_metadata_keys_to_lowercase_ascii() {
        assert_eq!(normalize_metadata_key("Authorization"), "authorization");
    }

    #[test]
    fn rejects_invalid_metadata_keys() {
        for key in ["", "Authorization", "trevrpc-timeout", "bad key"] {
            let mut metadata = Metadata::new();
            metadata.insert(key.to_owned(), Vec::new());

            let status = validate_metadata(&metadata).expect_err("metadata should be invalid");

            assert_eq!(status.code(), Code::InvalidArgument);
        }
    }

    #[test]
    fn rejects_metadata_limit_violations() {
        let mut too_many_entries = Metadata::new();
        for index in 0..=MAX_METADATA_ENTRIES {
            too_many_entries.insert(format!("key-{index}"), Vec::new());
        }
        assert!(validate_metadata(&too_many_entries).is_err());

        let mut value_too_large = Metadata::new();
        value_too_large.insert("key".to_owned(), vec![0; MAX_METADATA_VALUE_LEN + 1]);
        assert!(validate_metadata(&value_too_large).is_err());

        let mut total_too_large = Metadata::new();
        total_too_large.insert("key".to_owned(), vec![0; MAX_METADATA_TOTAL_SIZE + 1]);
        assert!(validate_metadata(&total_too_large).is_err());
    }
}
